### ISSUE TITLE
[bug] llamma server as a docker container seems to be locked to 1 CPU core

### ISSUE THREAD

**Issue by syangsao** [noonghunna/club-3090#90]
### What happened?

<img width="2924" height="1212" alt="Image" src="https://github.com/user-attachments/assets/4fe61933-6168-417d-9a6d-fbd24a332c6e" />

The llama-server seems to use only 1 CPU thread when running a long conversation.

### Steps to reproduce

```
1.  Kick off the installation with `bash scripts/launch.sh`
2.  Select single 3090
3.  Select option 1 
4.  Once installation completes, run the following to switch the LLM's `bash scripts/switch.sh --llamacpp/default`
5.  Kick off a long conversation
```

### Rig report — paste contents of `bash scripts/report.sh [--verify | --soak | --bench | --full]`

```markdown
# club-3090 rig report

Generated: 2026-05-06 23:04:24 UTC

_Redacted output (paths, host, user, tokens). Re-run with `--no-redact` for full data._

## System

- **OS:** Ubuntu 24.04.4 LTS
- **Kernel:** 6.17.0-22-generic
- **Environment:** bare metal
- **Locale:** en_US.UTF-8
- **Timezone:** CDT
- **Uptime:** up 2 days, 2 hours, 12 minutes

## CPU + RAM

- **CPU:** 11th Gen Intel(R) Core(TM) i9-11900K @ 3.50GHz (16 threads)
- **RAM:** 62Gi total, 47Gi available
- **Swap:** 8.0Gi

## Disk

- **~/club-3090/models-cache:** 540G available, ext4 filesystem
- **/var/lib/docker:** 540G available, ext4 filesystem

## GPU hardware

- **GPU 0:** NVIDIA GeForce RTX 3090 | 24576 MiB | driver 580.126.09 | VBIOS 94.02.42.00.A9 | persistence=Enabled
  - **Power:** limit=330.00 W (default=390.00 W, max=480.00 W) | current_draw=108.00 W ⚠ user-capped below default
  - **PCIe:** x16 lanes negotiated (GPU max x16, Gen up to 4) | bus 00000000:01:00.0
- **CUDA Runtime (per driver):** 13.0
- **ECC mode:** [N/A] (3090s don't have ECC; expect N/A)

### NVLink

_No NVLink detected (PCIe-only)_

### Topology

<details><summary>PCIe / GPU topology matrix</summary>


	GPU0	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	0-15	0		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks


</details>

### Full nvidia-smi

<details><summary>Full nvidia-smi output</summary>


Wed May  6 18:04:27 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.126.09             Driver Version: 580.126.09     CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3090        On  |   00000000:01:00.0 Off |                  N/A |
|  0%   29C    P2            108W /  330W |   20862MiB /  24576MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A          297315    C+G   ...c/gnome-remote-desktop-daemon        260MiB |
|    0   N/A  N/A          300464      G   /usr/lib/xorg/Xorg                       39MiB |
|    0   N/A  N/A          300559      G   /usr/bin/gnome-shell                     12MiB |
|    0   N/A  N/A          305620      C   /app/llama-server                     20500MiB |
+-----------------------------------------------------------------------------------------+


</details>

## Display / desktop state

- **$DISPLAY:** unset (headless)
- **Display processes running:** Xorg gnome-shell
- **GPU 0 idle VRAM:** 20862 MiB ⚠ something is using this GPU (display, browser, container)

## Container runtime

- **Docker:** 29.4.2
- **docker compose (v2):** 5.1.3
- **NVIDIA Container Toolkit:** 1.19.0

## Stack version

- **club-3090:** `6e57215` (branch: `master`)
- **GENESIS_PIN default:** `7b9fd319` (per scripts/setup.sh)

## Active container

_No vLLM container running. Start one with `bash scripts/launch.sh` and re-run for the full report._

---

_Generated by `bash scripts/report.sh`. Flags: `--verify` (verify-full), `--stress` (verify-stress 7/7 incl. Cliff 2 needles), `--soak` (SOAK_MODE=continuous, catches Cliff 2b), `--bench` (canonical TPS), `--full` (all four, ~35 min). Use `--no-redact` to disable redaction (internal sharing only)._
```

**Comment by noonghunna** [comment]
@syangsao — this is **expected behavior, not a bug**, but the framing is genuinely worth knowing for anyone reading this. Let me explain what you're seeing + give you a way to confirm whether decode is actually slow vs just looking idle.

## Why 1 CPU thread is correct with `-ngl 99` full GPU offload

Looking at our [`docker-compose.yml`](https://github.com/noonghunna/club-3090/blob/master/models/qwen3.6-27b/llama-cpp/compose/docker-compose.yml) for `llamacpp/default`, the relevant flags are:

- `-ngl 99` → push all 65 layers to the GPU
- `-fa on` → flash attention (GPU)
- `--cache-type-k q4_0 --cache-type-v q4_0` → KV cache lives on the GPU
- `-np 1` → single parallel slot

Under this config:
- **All matmul + attention compute runs on the GPU.** The 3090's tensor cores are the workhorse.
- **CPU's job is the coordination loop**: tokenizer → schedule a step → submit to GPU → receive logits → sample next token → stream out via SSE → repeat. That's ~1 thread of work, regardless of how long the conversation gets.

The "16 threads" your i9-11900K provides genuinely don't have heavy lifting to do during decode. llama.cpp **does** thread out prompt-processing chunks under high concurrency or batch=1 with a long initial prompt, but with our default `-np 1` + steady-state decode the CPU stays light.

This is structurally different from an OpenMP-style CPU-only inference run (`-ngl 0`) where you'd see all 16 threads pegged. Same image, totally different CPU profile depending on `-ngl`.

## How to confirm whether decode is *actually* slow

Your 330W earlier sweep had ~36 TPS narr/code. If a long conversation is producing materially less than that, **it's not a CPU thread issue** — it's something else. Two real culprits to check:

### 1. Long-conversation prompt re-processing

Every turn in a multi-turn conversation, llama.cpp re-processes the entire accumulated prompt. For a 50K-token conversation, that's 50K tokens of prefill work each turn. Prefill on Q3_K_XL on a 3090 runs ~1500 t/s, so 50K tokens = ~33 second wait per turn before generation starts. **This is what looks like "lag" but isn't a CPU bottleneck — it's prompt size.**

A cheap test: `time curl -sf -m 600 http://localhost:8020/v1/chat/completions -H 'Content-Type: application/json' -d '{"model":"qwen3.6-27b","messages":[{"role":"user","content":"Hi"}],"max_tokens":10}'` against a fresh server vs against your already-loaded long conversation. The fresh-server case will be fast (~1s); a multi-turn case will scale with prompt size.

### 2. Display contention on GPU 0

Your `report.sh` shows GPU 0 has these processes alongside `llama-server`:

```
gnome-remote-desktop-daemon         260MiB
/usr/lib/xorg/Xorg                   39MiB
/usr/bin/gnome-shell                 12MiB
```

That's ~310 MB of desktop/display rendering on the same GPU as inference. Contention is mild for VRAM but can hurt SM throughput when the desktop is doing anything (mouse, scrolling, redraw). If you SSH in headlessly with no Xorg, llama-server gets the whole GPU.

A clean test: temporarily stop the desktop (`sudo systemctl stop gdm` or similar) and re-time the same long conversation. If the slowness disappears, it was display contention, not CPU.

## Concrete diagnostic command

To get a clean TPS measurement against the running container, even mid-conversation:

```bash
git pull origin master    # you're on 6e57215, current master has script improvements
cd ~/club-3090
URL=http://localhost:8020 CONTAINER=llama-cpp-qwen36-27b MODEL=qwen3.6-27b \
  bash scripts/bench.sh
```

That'll run the canonical narr+code bench against your container and print wall TPS / decode TPS / TTFT. Compare to your earlier 36 TPS @ 330W baseline:

- If still ~36 TPS → decode is fine, what you're seeing is prompt-processing latency on long contexts. Expected with multi-turn workflows on llama.cpp at large ctx.
- If significantly lower → real perf regression worth investigating; file a follow-up with the bench numbers + a `nvidia-smi dmon -s pum` capture during the slow conversation.

## TL;DR

- 1 CPU thread on `-ngl 99` is correct, not a bug
- "Slowness in long conversations" usually = prompt re-processing scaling linearly with conversation size, not CPU-bound
- Run `bash scripts/bench.sh` against your container to differentiate "decode slow" from "prompt re-processing slow"
- Closing routing-wise as expected behavior. Reopen if `bench.sh` shows TPS materially below your 330W power-cap baseline (~36 TPS) — that'd be a real regression worth digging into.

Side note: while you're updating, master has a bunch of newer scripts now (engine-agnostic verify-full, auto-calibrating power-cap-sweep, etc.) — see [discussion #88](https://github.com/noonghunna/club-3090/discussions/88) for the contributor-script announcement.

**Comment by noonghunna** [comment]
Closing routing-wise — 1 CPU thread on -ngl 99 is expected behavior. Reopen if scripts/bench.sh shows TPS materially below the 330W baseline (~36 TPS).
